### ISSUE TITLE
Abstracting Dynamic Query API to It's Own Crate

### ISSUE THREAD

**Issue by zicklag** [jakobhellermann/bevy_mod_js_scripting#5]
I know this is super experimental, but I wanted to open this just to get my ideas out.

I was wondering whether or not it could make sense to move the dynamic query implementation to it's own crate that could be used in scripting plugins for any programming language.

For instance there's already a user asking about Lua ( #4 ) and in my project, while I do want to use JS/TS, I also want to make it run in the browser, so my binding needs are a little different than those provided in this crate, and I'll likely have to roll my own integration with Deno core and the browser.

If we abstracted the Bevy interface, though, it could allow all these different language integrations to re-use the dynamic world access layer which could be a big win.

**Comment by jakobhellermann** [comment]
That's definitely the goal, so far this project just served as a test bed for developing the necessary abstractions enabling dynamic scripting.

Out of curiosity, do you need just the query filter part (listing all entities matching a particular query) or would you also make use of the actual component pointer values?
So far I'm representing bevy values just by a base path into the world and a reflect path offset:

```rust
enum WorldBase {
    Component(Entity, ComponentId),
    Resource(ComponentId),
}

struct ValueRef {
    base: WorldBase,
    path: String,
}
```

I did this to get around issues like
```js
let component = world.get(my_entity, some_component_id);
world.despawn(my_entity);
print(component);
```
which you would have to deal with if you represent a component or resource values just as a pointer.

So every time you use the value in the scripting world it looks up the component pointer, so I don't even need the component pointer results from the query, just the entities are enough.

This means that this code
```js
let component = world.get(my_entity, some_component_id);
world.remove_component(my_entity, some_component_id);
world.insert_component(my_entity, some_component_id, value);
print(component); // still works here but references the new component value
```
works and prints the new value, which may be debatable.

**Comment by zicklag** [comment]
> Out of curiosity, do you need just the query filter part (listing all entities matching a particular query) or would you also make use of the actual component pointer values?

I haven't done any real-world testing yet, because I'm still working on integrating this with my game but I think the way you're handling it currently seems to make a lot of sense and seems intuitive to me.

And it seems a lot easier to implement safely.

**Comment by makspll** [comment]
I'll add my two cents since I also tried my hand at a scripting lib and got it to a usable state.
I also went down the Reflect Base + Reflect Path approach and that seems like the right way if we want to do things in safe Rust. I tried experimenting with caching path accesses, but that is a can of worms that I'd rather leave for later (was very difficult to do safely and makes the whole thing much more convoluted than it needs to be) + I am not convinced about much performance benefit.

The way you're structuring the API here seems spot on, and would likely directly plug into `bevy_mod_scripting` easilly. I actually don't use Component/ResourceID's yet I've found using `ReflectComponent` and `ReflectResource` directly pretty handy (although it's probably heavier on the cache lines) but using these ID's definitely seems like the right approach.

I would love to see a standalone crate for this!
